cdap-tms/src/main/java/io/cdap/cdap/messaging/service/CoreMessagingService.java

```diff
@@ -438,22 +438,27 @@ private Map<String, String> createDefaultProperties() {
     return properties;
   }
 
+  private TopicMetadata getTopic(TopicId topicId) throws TopicNotFoundException, IOException {
+    try {
+      return topicCache.get(topicId);
+    } catch (ExecutionException e) {
+      Throwable cause = Objects.firstNonNull(e.getCause(), e);
+      Throwables.propagateIfPossible(cause, TopicNotFoundException.class, IOException.class);
+      throw Throwables.propagate(e.getCause());
+    }
+  }
+
   @Override
   public CloseableIterator<RawMessage> fetch(MessageFetchRequest messageFetchRequest)
       throws TopicNotFoundException, IOException {
-    final TopicMetadata metadata =
-        new DefaultTopicMetadata(
-            messageFetchRequest.getTopicId(),
-            getTopicMetadataProperties(messageFetchRequest.getTopicId()));
-    MessageTable messageTable = createMessageTable(metadata);
+    TopicMetadata metadata = getTopic(messageFetchRequest.getTopicId());
+
+    TableProvider<MessageTable> messageTableProvider = () -> createMessageTable(metadata);
     try {
       return new MessageCloseableIterator(
-          metadata,
-          () -> createMessageTable(metadata),
-          () -> createPayloadTable(metadata),
-          messageFetchRequest);
+          metadata, messageTableProvider, () -> createPayloadTable(metadata), messageFetchRequest);
     } catch (Throwable t) {
-      closeQuietly(messageTable);
+      closeQuietly(messageTableProvider.get());
```

Comment:
This is  confusing. Why are we closing just one of those? (message table and not payload table)? I think removing the provider for message table would make things easier. You would just use `try with resource` for message table and keep payload as provider and let iterator handle it

Reply:
We were creating two message table, and always closing one of them. If there was an exception, the one created outside the iterator constructor was being closed. Otherwise, the one created inside iterator constructor was being closed.